### PR TITLE
Change the PHP project root detection method

### PR DESCRIPTION
**Pull request by lkebin**
Change method of find out project root directory.

For a PHP project, find out `composer.json` file to indicate the project root should be better than VCS folder.
Sometimes project use `git` but another using `hg` and `svn`, but if you are using php-language-server, you may also using composer.

### CODE REVIEW

**Comment by RyanSquared** [comment]
Is there a case where you're using Git but not using Composer? Is there any possibility that an ALE user might have their linter stop working because of this?

**Comment by lkebin** [comment]
@RyanSquared 
Yes, some PHP projects using Composer with [Mercurial (hg)](https://www.mercurial-scm.org/) or [SVN](https://subversion.apache.org/).
Yes. Some projects linter function will be stop working with this changes.

If composer.json is not a good idea for project root detection, may be replace the static file/directory to dynamic variable should be better.

**Comment by w0rp** [comment]
If some projects don't use composer, then support both. Check for composer.json first, and fall back on `.git`.

**Comment by lkebin** [comment]
Yes, we can support both of them. 

Maybe define a global variable will be better.  Such as 
```vim
g:ale_php_project_root_marker = ['.git', '.hg', 'composer.json', '.svn', '...']
```
Check each element until the first found, if nothing found `getcwd()` should be fine.
How do you think ?

**Comment by w0rp** [comment]
Yes, that might be a good option.

File: autoload/ale/path.vim [hunk 1]
```diff
@@ -17,6 +17,33 @@ function! ale#path#Simplify(path) abort
     return substitute(simplify(l:win_path), '^\\\+', '\', 'g') " no-custom-checks
 endfunction
 
+" Given a buffer and a file or directory name list, find the nearest file or directory by searching upwards
+" through the paths relative to the given buffer.
+function! ale#path#FindNearestMarker(buffer, markers) abort
+    let l:buffer_filename = fnamemodify(bufname(a:buffer), ':p')
+    let l:buffer_filename = fnameescape(l:buffer_filename)
+
+    for l:marker in a:markers
```
Comment (w0rp) [inline autoload/ale/path.vim]: Use a `for` loop with `ale#path#Upwards` instead. Have a look at the functions in the `python.vim` file for examples.

File: autoload/ale/path.vim [hunk 2]
```diff
@@ -129,8 +156,8 @@ endfunction
 function! ale#path#IsBufferPath(buffer, complex_filename) abort
     " If the path is one of many different names for stdin, we have a match.
     if a:complex_filename is# '-'
-    \|| a:complex_filename is# 'stdin'
-    \|| a:complex_filename[:0] is# '<'
+                \|| a:complex_filename is# 'stdin'
+                \|| a:complex_filename[:0] is# '<'
```
Comment (w0rp) [inline autoload/ale/path.vim]: Undo these changes.

File: autoload/ale/path.vim [hunk 3]
```diff
@@ -153,7 +180,7 @@ function! ale#path#IsBufferPath(buffer, complex_filename) abort
     let l:buffer_filename = expand('#' . a:buffer . ':p')
 
     return l:buffer_filename is# l:test_filename
-    \   || l:buffer_filename[-len(l:test_filename):] is# l:test_filename
+                \   || l:buffer_filename[-len(l:test_filename):] is# l:test_filename
```
Comment (w0rp) [inline autoload/ale/path.vim]: Undo these changes.

File: autoload/ale/path.vim [hunk 4]
```diff
@@ -187,13 +214,13 @@ function! ale#path#ToURI(path) abort
...
-    \   '/',
-    \   'g',
-    \)
+                \   ((l:has_drive_letter || a:path[:0] is# '/') ? 'file://' : '')
+                \       . (l:has_drive_letter ? '/' . a:path[:2] : '')
+                \       . ale#uri#Encode(l:has_drive_letter ? a:path[3:] : a:path),
+                \   '\\',
+                \   '/',
+                \   'g',
+                \)
```
Comment (w0rp) [inline autoload/ale/path.vim]: Undo these changes.

**Comment by lkebin** [comment]
@w0rp I did some changes according to your suggestions.

**Comment by lkebin** [comment]
Let me submit PR again when I have a free time to consider a better solution.
